pkt-line.c

```diff
@@ -152,19 +152,21 @@ static void format_packet(struct strbuf *out, const char *prefix,
 static int packet_write_fmt_1(int fd, int gently, const char *prefix,
 			      const char *fmt, va_list args)
 {
-	static struct strbuf buf = STRBUF_INIT;
+	struct strbuf buf = STRBUF_INIT;
```

Comment:
Could we follow the pattern of the later commits to have a `_r` version that takes a `struct strbuf` in the args, but then the current method still has a `static struct strbuf` so we don't have a performance regression? It just seems like we are adding a lot of allocations that previously did not exist, so that could cause performance issues in some cases.

The safest would be to have a "single-threaded, but fast" case and a "concurrent-safe, but each thread needs to manage its own strbuf" case.

Reply:
I concur about the allocations, and meant to add a TODO to look into this.

My main concern with these `*_r()` functions is: why are the non-`_r` versions not re-entrant? It's not like we're bound by some POSIX specifications to keep the original semantics.

I'd say: let's make the original functions re-entrant, and not add any extra `_r` functions to begin with.

And yes, handing down a `strbuf` makes the most sense, to avoid the cost of allocations.

Reply:
I was just trying to avoid touching all of the callers in the middle of what I was doing,
so I was trying to avoid breaking things -- also knowing how many conflicts I'd be creating
when I tried to push PC upstream.

An alternative approach would be to look at the pkt-line wrappers `packet_reader_*()` and `packet_writer_*()` that basically wrap the fd.  The reader side already has a buffer in the `struct packet_reader`.

Can I ask a larger question?   Do we actually need pkt-line as the protocol between the
fsmonitor client and the daemon?  If it is a single request with a few timestamp fields and
a basic multi-line response with a list of paths or something similar, [then] couldn't we
just use regular stdio of ascii/utf data and avoid all of these pkt-line issues.  This would
also give us the fwrite buffering for free.  Just a thought.  I haven't had time to study the
message formats in detail yet.

Reply:
> Do we actually need pkt-line as the protocol between the
> fsmonitor client and the daemon?

We don't _need_ it, but I thought it would be neat to have the built-in facilities like flush packets and byte-counted packets. It feels as if we would have to reinvent those if we don't use pkt-line.

Having said that, I'd be fine with just pulling out "pkt-line: use stack rather than static buffer in packet_write_gently()", shoving it into the next patch series, and dropping the "thread-safe-pkt-line" topic from the FSMonitor thicket.

Thoughts?

Reply:
I attempted to create a perf test in 93961d5381d3bd766275bb1dca338a358c9d5732 but could not measure any difference in the test helper:

```
Test                      v2.27.0.vfs.0.0      HEAD                    
-----------------------------------------------------------------------
0008.2: pkt-line pack     37.06(10.98+26.03)   37.39(11.04+26.15) +0.9%
0008.3: pkt-line unpack   27.67(12.55+15.12)   27.72(12.38+15.34) +0.2%
```

Reply:
That matches my expectations that the performance penalty of those allocations are dwarved by the remaining logic.

Also, the biggest cost of allocations come from vastly different allocation sizes (causing fragmentation) and multi-threaded allocations (which typically cause locking contention and/or suboptimal memory usage), neither of which I would expect to matter (read: to be exercised) in the perf test.